### ISSUE TITLE
CreditRiskPipeline (Full Pipeline on a Column Subset)

### ISSUE THREAD

**Issue by RigelAlgebar** [RigelAlgebar/ml-model-git-lab#8]
## 🧑‍💻 Issue 3 – Subset 3 CreditRiskPipeline (Full Pipeline on a Column Subset)

** Build a full CreditRiskPipeline class for subset 3, reusing transformers **

You will work with these subset-3 columns:

```
SUBSET3_COLS = [
    "id",
    "member_id",
    # numeric – balances, utilization, inquiries
    "tot_cur_bal",
    "open_acc_6m",
    "open_il_6m",
    "open_il_12m",
    "open_il_24m",
    "mths_since_rcnt_il",
    "total_bal_il",
    "il_util",
    "open_rv_12m",
    "open_rv_24m",
    "max_bal_bc",
    "all_util",
    "total_rev_hi_lim",
    "inq_fi",
    "total_cu_tl",
    "inq_last_12m",
    # categorical / dates
    "earliest_cr_line",
    "initial_list_status",
    "last_pymnt_d",
    "next_pymnt_d",
    "last_credit_pull_d",
    "application_type",
    "verification_status_joint",
]
```

> Target column for modeling: **`loan_status`** (from the full dataframe).

---

### 🎯 Goal

Create a **high-level credit risk pipeline** for subset 3 that:

- Cleans & transforms:
  - numeric balance/utilization/inquiry features  
  - date & categorical features from subset 3  
- Reuses the style of numeric + categorical transformers (Issues 1 & 2)  
- Wraps everything as a `CreditRiskPipeline` class with:
  - `.fit(X, y)`
  - `.predict(X)`
  - `.predict_proba(X)`
  - `.evaluate(X, y, metric)`

---

### ✅ Tasks

- [ ] In `src/config.py`, define subset-3 column groups:

```
SUBSET3_NUMERIC_COLS = [
    "tot_cur_bal",
    "open_acc_6m",
    "open_il_6m",
    "open_il_12m",
    "open_il_24m",
    "mths_since_rcnt_il",
    "total_bal_il",
    "il_util",
    "open_rv_12m",
    "open_rv_24m",
    "max_bal_bc",
    "all_util",
    "total_rev_hi_lim",
    "inq_fi",
    "total_cu_tl",
    "inq_last_12m",
]

SUBSET3_CATEGORICAL_COLS = [
    "earliest_cr_line",
    "initial_list_status",
    "last_pymnt_d",
    "next_pymnt_d",
    "last_credit_pull_d",
    "application_type",
    "verification_status_joint",
]
```

- [ ] In `src/cleaning_dates.py`, implement pure date feature functions:

```
def parse_dates(df: pd.DataFrame, cols: list[str]) -> pd.DataFrame:
    # Convert given columns to datetime (coerce errors to NaT).
    ...

def add_credit_history_years(df: pd.DataFrame) -> pd.DataFrame:
    # Create "credit_history_years" using "earliest_cr_line" and "last_credit_pull_d"
    # (or "issue_d" if needed).
    ...
```

- [ ] In `src/pipelines.py`, implement a functional helper to build a preprocessor for subset 3:

```
from sklearn.compose import ColumnTransformer

def build_subset3_preprocessor():
    num_cols = SUBSET3_NUMERIC_COLS
    cat_cols = SUBSET3_CATEGORICAL_COLS

    # Reuse or mimic numeric/categorical transformers from Issues 1 & 2:
    num_transformer = Subset1NumericBorrowerTransformer(numeric_cols=num_cols)
    cat_transformer = Subset2CategoricalPerformanceTransformer(cat_cols=cat_cols)

    preprocessor = ColumnTransformer(
        transformers=[
            ("num", num_transformer, num_cols),
            ("cat", cat_transformer, cat_cols),
        ]
    )
    return preprocessor
```

- [ ] In `src/pipelines.py`, implement `CreditRiskPipeline` OOP class:

```
from sklearn.pipeline import Pipeline

class CreditRiskPipeline:
    def __init__(self, numeric_cols, categorical_cols, model):
        self.numeric_cols = numeric_cols
        self.categorical_cols = categorical_cols
        self.model = model
        self.pipeline = None

    def build_pipeline(self):
        preprocessor = build_subset3_preprocessor()
        self.pipeline = Pipeline(steps=[
            ("preprocess", preprocessor),
            ("model", self.model),
        ])

    def fit(self, X, y):
        if self.pipeline is None:
            self.build_pipeline()
        self.pipeline.fit(X, y)
        return self

    def predict(self, X):
        return self.pipeline.predict(X)

    def predict_proba(self, X):
        return self.pipeline.predict_proba(X)

    def evaluate(self, X, y, metric):
        y_pred = self.predict(X)
        return metric(y, y_pred)
```

- [ ] Add a demo notebook/script using subset-3 features:

```
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier

from src.config import SUBSET3_NUMERIC_COLS, SUBSET3_CATEGORICAL_COLS
from src.pipelines import CreditRiskPipeline

feature_cols = SUBSET3_NUMERIC_COLS + SUBSET3_CATEGORICAL_COLS
X = df[feature_cols]
y = df["loan_status"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

model = RandomForestClassifier(n_estimators=200, random_state=42)
crp = CreditRiskPipeline(SUBSET3_NUMERIC_COLS, SUBSET3_CATEGORICAL_COLS, model)
crp.fit(X_train, y_train)
acc = crp.evaluate(X_test, y_test, accuracy_score)
print("Subset 3 – Test accuracy:", acc)
```

- [ ] Add tests in `tests/test_credit_risk_pipeline_subset3.py`:

  - `fit` runs end-to-end on a small sample  
  - `predict` returns an array of correct length  
  - `evaluate` returns a scalar metric (e.g. accuracy)

---

### ✅ Acceptance Criteria

- Subset-3 numeric & categorical/date features go through a single end-to-end pipeline  
- Implementation uses **functional helpers** (`build_subset3_preprocessor`, date functions) + **`CreditRiskPipeline` OOP**  
- Tests confirm that the pipeline can be trained, predicts, and evaluates on subset-3 data
